### PR TITLE
List command of plugin

### LINKED ISSUE #320: List plugin
- parent: #316 
- depends: #298 
---

`~$ vatz plugin list` command will show list of installed plugin info.
Ref. https://github.com/dsrvlabs/vatz/pull/317#issuecomment-1334749495

### PR DESCRIPTION
**Pull request by rootwarp**
### 1. Type of change

Please delete options that are not relevant.

- [x] New feature
- [ ] Enhancement
- [ ] Bug/fix (non-breaking change which fixes an issue)
- [ ] others (anything other than above)

---

**Related:** # (issue)

close #320 

**Summary**

List command of plugin that are installed on local machine.

```
~$ ./vatz plugin install https://github.com/dsrvlabs/vatz-plugin-cosmoshub/plugins/node_active_status cosmos-status

~$ vatz plugin list
```

---

### CODE REVIEW

**Comment by rootwarp** [comment]
@xellos00 Please take look at this before merging #317.

**Comment by xellos00** [comment]
> @xellos00 Please take look at this before merging #317. I'll rebase this after #317 merged.

so review this first, before merge #317 right?

**Comment by rootwarp** [comment]
> > @xellos00 Please take look at this before merging #317. I'll rebase this after #317 merged.
> 
> so review this first, before merge #317 right?

@xellos00 
Yes. Due to the previous PR being delayed to merging, I implemented this feature based on previous PR in advance. I just hope that you can inspect this code in detail and get more feedback for new feature.

**Comment by xellos00** [comment]
> > > @xellos00 Please take look at this before merging #317. I'll rebase this after #317 merged.
> > 
> > 
> > so review this first, before merge #317 right?
> 
> @xellos00 Yes. Due to the previous PR being delayed to merging, I implemented this feature based on previous PR in advance. I just hope that you can inspect this code in detail and get more feedback for new feature.

got it!!

**Comment by xellos00** [comment]
```
 dongyookang DK ⚡️   ~/ffplay/GolandProjects/dsrvlabs/vatz   list
 make coverage
fatal: No names found, cannot describe anything.
echo "Test Coverage script will be here"
Test Coverage script will be here
# github.com/dsrvlabs/vatz/manager/plugin.test
ld: warning: -no_pie is deprecated when targeting new OS versions
?   	github.com/dsrvlabs/vatz	[no test files]
# github.com/dsrvlabs/vatz/cmd.test
ld: warning: -no_pie is deprecated when targeting new OS versions
ok  	github.com/dsrvlabs/vatz/cmd	2.711s	coverage: 17.1% of statements
ok  	github.com/dsrvlabs/vatz/manager/api	0.378s	coverage: 0.0% of statements [no tests to run]
ok  	github.com/dsrvlabs/vatz/manager/config	0.444s	coverage: 91.2% of statements
ok  	github.com/dsrvlabs/vatz/manager/dispatcher	0.464s	coverage: 7.3% of statements
ok  	github.com/dsrvlabs/vatz/manager/executor	0.557s	coverage: 68.9% of statements
ok  	github.com/dsrvlabs/vatz/manager/healthcheck	0.672s	coverage: 46.3% of statements
ok  	github.com/dsrvlabs/vatz/manager/plugin	2.877s	coverage: 68.0% of statements
?   	github.com/dsrvlabs/vatz/manager/types	[no test files]
?   	github.com/dsrvlabs/vatz/mocks	[no test files]
ok  	github.com/dsrvlabs/vatz/rpc	1.792s	coverage: 67.2% of statements
ok  	github.com/dsrvlabs/vatz/utils	0.163s	coverage: 50.0% of statements
 dongyookang DK ⚡️   ~/ffplay/GolandProjects/dsrvlabs/vatz   list
```

**Comment by rootwarp** [comment]
@xellos00  Responses.

> 
> @rootwarp Can you Check up followings?
> 
> 1. Column name typo
>    
>    * INSTALL DATA  => INSTALLATION DATE

I'll change it.

> 2. Column Order
>    
>    * Name, REPOSITORY, VERSION, INSTALLATION DATE

Also I'll change it.

> 3. Get rid of following lines?
>    Or is this intentionally printed out?
> 
> ```
>  ./vatz plugin list
> 2022-12-12T13:17:12-06:00 INF List plugins module=plugin
> 2022-12-12T13:17:12-06:00 INF List module=plugin
> 2022-12-12T13:17:12-06:00 INF newReader /Users/dongyookang/.vatz/vatz.db module=db
> 2022-12-12T13:17:12-06:00 INF Create DB Instance module=db
> 2022-12-12T13:17:12-06:00 INF List Plugin module=db
> ```

Regarding this, I thought few hours before implementing this and I think we need to discuss and decide policies for logs.

As you can see, this problem is caused because the output channels for the log and the result are identical.

We may remove the logging code or disable log outputs for some command lines
and these could be the item that we have to discuss.
Before we discuss this, I think it could be better to keep log messages that will help implementation.

File: cmd/plugin.go [hunk 1]
```diff
@@ -108,6 +109,37 @@ var (
...
+				log.Error().Str("module", "plugin").Err(err)
+				return err
+			}
+
+			w := table.NewWriter()
+			w.SetOutputMirror(os.Stdout)
+			w.AppendHeader(table.Row{"Name", "Install Data", "Repository", "Version"})
+
+			for _, plugin := range plugins {
+				dateStr := plugin.InstalledAt.Format("2006-01-02 15:04:05")
```
Comment (skonhwang) [inline cmd/plugin.go:132]: 2006-01-02 15:04:05 ??

Comment (rootwarp) [reply cmd/plugin.go:132]: @skonhwang This is unique style of Golang's date formatting.

Ref. https://pkg.go.dev/time#pkg-constants

Comment (skonhwang) [reply cmd/plugin.go:132]: @rootwarp , Thanks to you I am more smarter about gland's   format of date

File: cmd/plugin.go [hunk 2]
```diff
@@ -108,6 +109,37 @@ var (
 			return mgr.Start(pluginName, exeArgs)
 		},
 	}
+
+	listCommand = &cobra.Command{
+		Use:     "list",
+		Short:   "List installed plugin",
+		Example: "vats plugin list",
```
Comment (meetrick) [inline cmd/plugin.go:116]: vats -> vatz

File: manager/plugin/db.go [hunk 3]
```diff
@@ -122,12 +131,36 @@ func (p *pluginDB) DeletePlugin(name string) error {
 
 func (p *pluginDB) UpdatePlugin() error {
 	// TODO:
+
```
Comment (meetrick) [inline manager/plugin/db.go:134]: Unnecessary modification

**Review by skonhwang — APPROVED** [review]

**Review by meetrick — COMMENTED** [review]
I haven't tested it yet, but there seems to be no problem with the operation.
But please check typo error comment.

**Review by heejin-github — APPROVED** [review]
LGTM!

File: cmd/plugin.go (hunk 2 above)
Comment (xellos00) [reply cmd/plugin.go:116]: is this updated?

File: manager/plugin/db.go (hunk 3 above)
Comment (rootwarp) [reply manager/plugin/db.go]: @meetrick removed thx.

File: cmd/plugin.go (hunk 2 above)
Comment (rootwarp) [reply cmd/plugin.go]: @meetrick @xellos00 Fixed. thx.

**Pull request merged**